plugins/outputs/azure_monitor/azure_monitor.go

```diff
@@ -0,0 +1,491 @@
+	h.Write(b[:n])
+	h.Write([]byte("\n"))
+	return h.Sum64()
+}
+
+func translate(m telegraf.Metric, prefix string) *azureMonitorMetric {
+	var dimensionNames []string
+	var dimensionValues []string
+	for i, tag := range m.TagList() {
+		// Azure custom metrics service supports up to 10 dimensions
+		if i > 10 {
+			log.Printf("W! [outputs.azure_monitor] metric [%s] exceeds 10 dimensions", m.Name())
+			continue
+		}
+		dimensionNames = append(dimensionNames, tag.Key)
+		dimensionValues = append(dimensionValues, tag.Value)
+	}
+
+	min, _ := m.GetField("min")
+	max, _ := m.GetField("max")
+	sum, _ := m.GetField("sum")
+	count, _ := m.GetField("count")
+	return &azureMonitorMetric{
+		Time: m.Time(),
+		Data: &azureMonitorData{
+			BaseData: &azureMonitorBaseData{
+				Metric:         strings.SplitN(m.Name(), "-", 2)[1],
+				Namespace:      prefix + strings.SplitN(m.Name(), "-", 2)[0],
+				DimensionNames: dimensionNames,
+				Series: []*azureMonitorSeries{
+					&azureMonitorSeries{
+						DimensionValues: dimensionValues,
+						Min:             min.(float64),
+						Max:             max.(float64),
+						Sum:             sum.(float64),
+						Count:           count.(int64),
+					},
+				},
+			},
+		},
+	}
+}
+
+// Add will append a metric to the output aggregate
+func (a *AzureMonitor) Add(m telegraf.Metric) {
+	// Azure Monitor only supports aggregates 30 minutes into the past and 4
+	// minutes into the future. Future metrics are dropped when pushed.
+	t := m.Time()
+	tbucket := time.Date(t.Year(), t.Month(), t.Day(), t.Hour(), t.Minute(), 0, 0, t.Location())
+	if tbucket.Before(time.Now().Add(-time.Minute * 30)) {
+		log.Printf("W! attempted to aggregate metric over 30 minutes old: %v, %v", t, tbucket)
+		return
+	}
+
+	// Azure Monitor doesn't have a string value type, so convert string fields
+	// to dimensions (a.k.a. tags) if enabled.
+	if a.StringsAsDimensions {
+		for fk, fv := range m.Fields() {
+			if v, ok := fv.(string); ok {
+				m.AddTag(fk, v)
+			}
+		}
+	}
+
+	for _, f := range m.FieldList() {
+		fv, ok := convert(f.Value)
+		if !ok {
+			continue
+		}
+
+		// Azure Monitor does not support fields so the field name is appended
+		// to the metric name.
+		name := m.Name() + "-" + sanitize(f.Key)
+		id := hashIDWithField(m.HashID(), f.Key)
+
+		_, ok = a.cache[tbucket]
+		if !ok {
+			// Time bucket does not exist and needs to be created.
+			a.cache[tbucket] = make(map[uint64]*aggregate)
+		}
+
+		nf := make(map[string]interface{}, 4)
+		nf["min"] = fv
+		nf["max"] = fv
+		nf["sum"] = fv
+		nf["count"] = 1
+		// Fetch existing aggregate
+		agg, ok := a.cache[tbucket][id]
+		if ok {
+			aggfields := agg.Fields()
+			if fv > aggfields["min"].(float64) {
+				nf["min"] = aggfields["min"]
+			}
+			if fv < aggfields["max"].(float64) {
+				nf["max"] = aggfields["max"]
+			}
+			nf["sum"] = fv + aggfields["sum"].(float64)
+			nf["count"] = aggfields["count"].(int64) + 1
+		}
+
+		na, _ := metric.New(name, m.Tags(), nf, tbucket)
```

Comment:
https://github.com/influxdata/telegraf/pull/4441

Reply:
`telegraf.Metric` was required to separate the aggregation step and the write step to utilize Telegraf's internal buffering, via https://github.com/influxdata/telegraf/commit/19e435fc618b25a66b0cb935c37199e452aca1a8. I don't see a way to use a new type and still use the internal buffering. Let me know if you have any ideas. This could also be left as an optimization for a later time.

Reply:
The metric would come in via Add as a telegraf.Metric, you would convert it to a custom type for the cache, and then it would be turned back into a telegraf.Metric in the Push step.  It seems like a lot of conversions but you are actually already making a new cache Metric for each telegraf.Metric added.  However, you don't really want or need the loose field typing of a telegraf.Metric or any of the other functionality.  This would allow you to remove all the conversions above.